### PR TITLE
properly update size in randomEvicted cache when replacing entries

### PR DESCRIPTION
**Pull request by songgao**

### CODE REVIEW

File: cache/cache.go [hunk 1]
```diff
@@ -78,9 +78,11 @@ func (c *randomEvictedCache) Add(key string, data Measurable) {
 	}
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	if _, ok := c.data[key]; !ok {
-		c.cachedBytes += increase
+	if v, ok := c.data[key]; ok {
+		decrease := 2*len(key) + v.Size()
```
Comment (jzila) [inline cache/cache.go]: It seems like the pattern `2*len(key) + val.Size()` is used a bunch: should it be a function?

Comment (songgao) [inline cache/cache.go]: sure; will do

File: cache/cache.go [hunk 2]
```diff
@@ -48,12 +48,22 @@ func NewRandomEvictedCache(maxBytes int) Cache {
 	}
 }
 
+func (c *randomEvictedCache) entrySize(key *string, value Measurable) int {
```
Comment (jzila) [inline cache/cache.go]: Why is the string a pointer? Strings are copy-on-write, aren't they?

Comment (songgao) [inline cache/cache.go]: Really? I had no idea! I thought it was always copied

Comment (songgao) [inline cache/cache.go]: Updated; thanks!

Comment (jzila) [inline cache/cache.go]: Looks like calling it copy-on-write is not strictly accurate, even if the semantics are the same:
https://stackoverflow.com/a/8532749

Comment (songgao) [inline cache/cache.go]: Yeah it's immutable, so there's no "write". But for the same reason it's not copied, so still good point!

**Pull request merged**